backend/src/main/java/ai/verta/modeldb/experimentRun/subtypes/ObservationHandler.java

```diff
@@ -228,34 +232,70 @@ public ObservationHandler(Executor executor, FutureJdbi jdbi) {
         handle -> {
-          // Delete from keyvalue
-          var sql =
-              "delete from keyvalue where id in "
-                  + "(select keyvaluemapping_id from observation where entity_name=:entity_name and field_type=:field_type and experiment_run_id=:run_id)";
+          handle.useTransaction(
+              TransactionIsolationLevel.SERIALIZABLE,
+              handle1 -> {
+                // Delete from keyvalue
+                var fetchQueryString = "select o.id, o.keyvaluemapping_id from observation as o ";
+                if (maybeKeys.isPresent()) {
+                  fetchQueryString += " INNER JOIN keyvalue as kv ON o.keyvaluemapping_id = kv.id ";
+                }
+                fetchQueryString +=
+                    " WHERE o.experiment_run_id = :run_id " + " AND o.entity_name = :entityName";
 
-          if (maybeKeys.isPresent()) {
-            sql += " and kv_key in (<keys>)";
-          }
+                if (maybeKeys.isPresent()) {
+                  fetchQueryString += " AND kv.kv_key IN (<keys>)";
+                }
+                var query =
+                    handle
+                        .createQuery(fetchQueryString)
+                        .bind("run_id", runId)
+                        .bind("entityName", "ExperimentRunEntity");
 
-          var query =
-              handle
-                  .createUpdate(sql)
-                  .bind("entity_name", "\"ExperimentRunEntity\"")
-                  .bind("field_type", "\"observations\"")
-                  .bind("run_id", runId);
+                if (maybeKeys.isPresent()) {
+                  query = query.bindList("keys", maybeKeys.get());
+                }
+                var observationKVMappingList =
+                    query
+                        .map(
+                            (rs, ctx) ->
+                                new AbstractMap.SimpleEntry<>(
+                                    rs.getLong("id"), rs.getLong("keyvaluemapping_id")))
+                        .list();
 
-          if (maybeKeys.isPresent()) {
-            query = query.bindList("keys", maybeKeys.get());
-          }
+                // Remove foreignKey constraint first
+                handle
+                    .createUpdate(
+                        "update observation set keyvaluemapping_id = null where id in (<ob_ids>)")
+                    .bindList(
+                        "ob_ids",
+                        observationKVMappingList.stream()
+                            .map(AbstractMap.SimpleEntry::getKey)
+                            .collect(Collectors.toList()))
+                    .execute();
 
-          query.execute();
+                // Delete KeyValue mapped with Observation by Ids
+                handle
+                    .createUpdate("delete from keyvalue where id in (<kv_ids>)")
+                    .bind("entity_name", "ExperimentRunEntity")
+                    .bind("field_type", "observations")
```

Comment:
Same

Reply:
this is not new code it is restructured from https://github.com/VertaAI/modeldb/pull/2285/files#diff-8c31b60eb34e831c4c2df7dcf7ca2be57abc9ff2dc69bbaaa83eb6d9f4ae2f34L243